bundles/org.openhab.binding.gpio/src/main/java/org/openhab/binding/gpio/internal/handler/PigpioDigitalOutputHandler.java

```diff
@@ -52,34 +63,182 @@ public class PigpioDigitalOutputHandler implements ChannelHandler {
+                        job.cancel(false);
+                    }
+
+                    this.pulseJob = scheduler.schedule(() -> handlePulseCommand(command), this.pulseTimeout,
+                            TimeUnit.MILLISECONDS);
+                }
             }
         }
-        if (command instanceof OnOffType) {
-            try {
-                gpio.setValue(configuration.invert != (OnOffType.ON.equals(command)));
-            } catch (PigpioException e) {
-                logger.warn("An error occured while changing the gpio value: {}", e.getMessage());
+    }
+
+    public void handlePulseCommand(@Nullable Command command) {
+        OnOffType eCommand = OnOffType.OFF;
+
+        try {
+            synchronized (handleLock) {
+                GPIO lgpio = this.gpio;
+                Consumer<State> lupdateStatus = this.updateStatus;
+                Future<?> job = this.pulseJob;
+
+                if (lgpio == null) {
+                    return;
+                }
+
+                if (command instanceof OnOffType) {
+                    if (this.pulseCommand != null) {
+                        if ("ON".equals(this.pulseCommand)) {
+                            eCommand = OnOffType.ON;
+                        } else if ("OFF".equals(this.pulseCommand)) {
+                            eCommand = OnOffType.OFF;
+                        } else if ("BLINK".equals(this.pulseCommand)) {
+                            if (OnOffType.ON.equals(command)) {
+                                eCommand = OnOffType.OFF;
+                            } else if (OnOffType.OFF.equals(command)) {
+                                eCommand = OnOffType.ON;
+                            }
+                        }
+                    } else {
+                        if (OnOffType.ON.equals(command)) {
+                            eCommand = OnOffType.OFF;
+                        } else if (OnOffType.OFF.equals(command)) {
+                            eCommand = OnOffType.ON;
+                        }
+                    }
+
+                    logger.debug("gpio pulse command : {} {}", this.gpioId, eCommand.toString());
+
+                    lgpio.setValue(configuration.invert != (OnOffType.ON.equals(eCommand)));
+                    if (lupdateStatus != null) {
+                        lupdateStatus.accept((State) eCommand);
+                    }
+
+                    lastPulseCommand = eCommand;
+
+                    if ("BLINK".equals(this.pulseCommand) && this.pulseTimeout > 0) {
+                        final OnOffType feCommand = eCommand;
+                        if (job != null) {
+                            job.cancel(false);
+                        }
+                        this.pulseJob = scheduler.schedule(() -> handlePulseCommand(feCommand), this.pulseTimeout,
+                                TimeUnit.MILLISECONDS);
+                    }
+                }
+            }
+        } catch (Exception e) {
```

Comment:
Would it be possible to make the exception more specific? And would it be needed to set the thing offline if sending the command failed?

Reply:
All most alle was addressed except this in one. I just comment so that it is not overlooked

Reply:
Hmmm, the handlePulseCommand is always in response to a handleCommand (IE, a command issued by openHab). A failed handleCommand will result in the thing going offline as the exception is caught in PigpioRemoteHandler. But handlePulseCommand is in the subservient PigpioDigitalOutputHandler and executed from a timer, so I cannot easily propagate the exception.

I guess going offline from down that far would require a refactoring of the class structure and code. I never gave it much thought as the feature was a late addition.

But I may suggest wet it go for a few reasons:

1.  The fact that the failed pulse command is in response to a real command. Any subsequent real command will offline the thing. 
2. If the thing is really offline. The connect poll will offline the thing as part of the failed poll cycle (and subsequently reconnect it).

I did update the message to something hopefully more understandable:

`
        } catch (Exception e) {
            logger.warn(
                    "Pulse command exception, {} command may not have been received by pigpiod resulting in an unknown state:",
                    eCommand.toString(), e);
        }
`

Reply:
Good enough if you ask me.